src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,1758 @@
+        spirv_header.push_back(ds.storage_buffer_count);
+        spirv_header.push_back(padded_str_length);
+        internal_assert(padded_name.size() == padded_str_length);
+        spirv_header.insert(spirv_header.end(), (const uint32_t *)padded_name.data(), (const uint32_t *)(padded_name.data() + padded_name.size()));
+        ++index;
+    }
+    uint32_t header_word_count = spirv_header.size();
+    spirv_header.insert(spirv_header.begin(), header_word_count + 1);
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::compile(std::vector<char> &module) {
+    debug(2) << "CodeGen_Vulkan_Dev::SPIRV_Emitter::compile\n";
+
+    // First encode the descriptor set bindings for each entry point
+    // as a sidecar which we will add as a preamble header to the actual
+    // SPIR-V binary so the runtime can know which descriptor set to use
+    // for each entry point
+    SpvBinary spirv_header;
+    encode_header(spirv_header);
+
+    // Finalize and encode the SPIR-V IR into a compliant binary
+    SpvBinary spirv_binary;
+    builder.finalize();
+    builder.encode(spirv_binary);
+
+    size_t header_bytes = spirv_header.size() * sizeof(uint32_t);
+    size_t binary_bytes = spirv_binary.size() * sizeof(uint32_t);
+
+    debug(2) << "    encoding module ("
+             << "header_size: " << (uint32_t)(header_bytes) << ", "
+             << "binary_size: " << (uint32_t)(binary_bytes) << ")\n";
+
+    // Combine the header and binary into the module
+    module.reserve(header_bytes + binary_bytes);
+    module.insert(module.end(), (const char *)spirv_header.data(), (const char *)(spirv_header.data() + spirv_header.size()));
+    module.insert(module.end(), (const char *)spirv_binary.data(), (const char *)(spirv_binary.data() + spirv_binary.size()));
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::add_kernel(const Stmt &s,
+                                                   const std::string &name,
+                                                   const std::vector<DeviceArgument> &args) {
+    debug(2) << "Adding Vulkan kernel " << name << "\n";
+
+    // Add function definition
+    // TODO: can we use one of the function control annotations?
+
+    // We'll discover the workgroup size as we traverse the kernel
+    workgroup_size[0] = 0;
+    workgroup_size[1] = 0;
+    workgroup_size[2] = 0;
+
+    // Keep track of the descriptor set needed to bind this kernel's inputs / outputs
+    DescriptorSet descriptor_set;
+    descriptor_set.entry_point_name = name;
+
+    // Declare the kernel function
+    SpvId void_type_id = builder.declare_void_type();
+    SpvId kernel_func_id = builder.add_function(name, void_type_id);
+    SpvFunction kernel_func = builder.lookup_function(kernel_func_id);
+    builder.enter_function(kernel_func);
+
+    // TODO: only add the SIMT intrinsics used
+    SpvFactory::Variables entry_point_variables;
+    auto intrinsics = {"WorkgroupId", "LocalInvocationId"};
+    for (const std::string &intrinsic_name : intrinsics) {
```

Comment:
Nit: `intrinsics` is `const char*[]`, so this does an implicit copy.  Probably better to either make `intrinsics` `std::string []` or to use `const char* const&` in the loop

Reply:
Nice catch!  I'll add this to my next batch of fixes.